# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1753826358 -7200
# Parent  1087d5d1817d3acb4288b4b85a655a03294d374d
Feature: Re-implementation of the two-line view with columns for BB 115.
* * *
Feature: Re-implementation of the two-line view: Make correspondents column pixel perfect.
* * *
Feature: Reduce row height in multi-line view to match height of cards.
* * *
Misc: Reduce thread indentation from 16px to 10px.

diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -375,8 +375,9 @@ pref("mail.showCondensedAddresses", true
 pref("mailnews.attachments.display.start_expanded", false);
 pref("mailnews.attachments.display.top", false);
 // hidden pref for changing how we present attachments in the message pane
 pref("mail.pane_config.dynamic", 0);
+pref("mail.pane_config.multiline_all", false);
 pref("editor.singleLine.pasteNewlines", 4);  // substitute commas for new lines in single line text boxes
 pref("editor.CR_creates_new_p", true);
 pref("mail.compose.default_to_paragraph", true);
 
diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -4740,8 +4740,9 @@ var threadPane = {
     this.setUpTagStyles();
     Services.prefs.addObserver("mailnews.tags.", this);
     Services.prefs.addObserver("mail.threadpane.table.horizontal_scroll", this);
     Services.prefs.addObserver("mail.threadpane.listview", this);
+    Services.prefs.addObserver("mail.pane_config.multiline_all", this);
 
     Services.obs.addObserver(this, "addrbook-displayname-changed");
     Services.obs.addObserver(this, "custom-column-added");
     Services.obs.addObserver(this, "custom-column-removed");
@@ -4791,8 +4792,15 @@ var threadPane = {
       () => this.updateThreadItemSize(),
       prefVal => Math.min(Math.max(2, prefVal), 3)
     );
 
+    // Setup multi-line.
+    const multiline = Services.prefs.getBoolPref(
+      "mail.pane_config.multiline_all",
+      false
+    );
+    customElements.get("thread-row").MULTILINE = multiline;
+
     window.addEventListener("uidensitychange", () => {
       this.updateThreadItemSize();
     });
     window.addEventListener("uifontsizechange", () => {
@@ -4977,8 +4985,18 @@ var threadPane = {
           }
           break;
         }
 
+        if (data == "mail.pane_config.multiline_all") {
+          const multiline = Services.prefs.getBoolPref(
+            "mail.pane_config.multiline_all",
+            false
+          );
+          customElements.get("thread-row").MULTILINE = multiline;
+          this.updateThreadItemSize();
+          break;
+        }
+
         if (data.startsWith("mailnews.tags.")) {
           this.setUpTagStyles();
           break;
         }
@@ -5916,33 +5934,41 @@ var threadPane = {
     let rowHeight = Math.ceil(currentFontSize * 1.4);
     let lineGap;
     let densityPaddingConstant;
     let cardRowHeight;
+    let multiLineFactor;
     switch (UIDensity.prefValue) {
       case UIDensity.MODE_COMPACT:
         // Calculation based on card components:
         lineGap = 1;
         densityPaddingConstant = 3; // card padding-block + 2 * row padding-block
         cardRowHeight =
           cardRowConstant + lineGap * this.rowCount + densityPaddingConstant;
+        multiLineFactor = 1.7;
         break;
       case UIDensity.MODE_TOUCH:
         rowHeight = rowHeight + 13;
         lineGap = 6;
         densityPaddingConstant = 12; // card padding-block + 2 * row padding-block
         cardRowHeight =
           cardRowConstant + lineGap * this.rowCount + densityPaddingConstant;
+        multiLineFactor = 1.6;
         break;
       default:
         rowHeight = rowHeight + 7;
         lineGap = 3;
         densityPaddingConstant = 7; // card padding-block + 2 * row padding-block
         cardRowHeight =
           cardRowConstant + lineGap * this.rowCount + densityPaddingConstant;
+        multiLineFactor = 1.7;
         break;
     }
     cardClass.ROW_HEIGHT = Math.max(cardRowHeight, 40);
-    rowClass.ROW_HEIGHT = Math.max(rowHeight, 18);
+    if (rowClass.MULTILINE) {
+      rowClass.ROW_HEIGHT = Math.max(rowHeight, 20) * multiLineFactor;
+    } else {
+      rowClass.ROW_HEIGHT = Math.max(rowHeight, 20);
+    }
   },
 
   /**
    * Update thread item size in DOM (thread cards and rows).
@@ -6218,8 +6244,12 @@ var threadPane = {
    */
   applyPersistedColumnsState(columnStates) {
     this.columns.forEach(c => {
       c.hidden = !columnStates[c.id]?.visible;
+      // Clear traces of old multi-line via hidden subject.
+      if (c.id == "subjectCol") {
+        c.hidden = false;
+      }
       c.ordinal = columnStates[c.id]?.ordinal ?? 0;
     });
     // Sort columns by ordinal.
     this.columns.sort(function (a, b) {
diff --git a/mail/base/content/about3Pane.xhtml b/mail/base/content/about3Pane.xhtml
--- a/mail/base/content/about3Pane.xhtml
+++ b/mail/base/content/about3Pane.xhtml
@@ -229,8 +229,16 @@
       </td>
       <td class="attachmentcol-column button-column" data-column-name="attachmentcol">
         <img src="" data-l10n-id="tree-list-view-row-attach" />
       </td>
+      <td class="tree-view-row-unread unreadbuttoncolheader-column button-column" data-column-name="unreadcol">
+        <button type="button"
+                class="button-flat tree-button-unread"
+                aria-hidden="true"
+                tabindex="-1">
+          <img src="" alt="" />
+        </button>
+      </td>
       <td class="subjectcol-column" data-column-name="subjectcol">
         <div class="thread-container">
           <button type="button"
                   class="button button-flat button-reset twisty"
@@ -242,16 +250,8 @@
             <img src="" alt="" /><span></span>
           </div>
         </div>
       </td>
-      <td class="tree-view-row-unread unreadbuttoncolheader-column button-column" data-column-name="unreadcol">
-        <button type="button"
-                class="button-flat tree-button-unread"
-                aria-hidden="true"
-                tabindex="-1">
-          <img src="" alt="" />
-        </button>
-      </td>
       <td class="sendercol-column" data-column-name="sendercol"></td>
       <td class="recipientcol-column" data-column-name="recipientcol"></td>
       <td class="correspondentcol-column" data-column-name="correspondentcol"></td>
       <td class="tree-view-row-spam junkstatuscol-column button-column" data-column-name="junkstatuscol">
diff --git a/mail/base/content/mailWindowOverlay.js b/mail/base/content/mailWindowOverlay.js
--- a/mail/base/content/mailWindowOverlay.js
+++ b/mail/base/content/mailWindowOverlay.js
@@ -469,8 +469,17 @@ function InitViewLayoutStyleMenu(event, 
     parent
       .querySelector(`[name="threadheader"]`)
       .setAttribute("checked", "true");
   }
+
+  const multiline = document.getElementById(
+    appmenu ? "appmenu_multilineAllFolders" : "multilineAllFolders"
+  );
+  if (Services.prefs.getBoolPref("mail.pane_config.multiline_all")) {
+    multiline.setAttribute("checked", "true");
+  } else {
+    multiline.removeAttribute("checked");
+  }
 }
 
 /**
  * Called when showing the menu_viewSortPopup menupopup, so it should always
@@ -1374,8 +1383,23 @@ function MsgFilters(emailAddress, folder
     MsgFilterList(args);
   }
 }
 
+function ToogleMultilineAll(appmenu) {
+  const multiline = document.getElementById(
+    appmenu ? "appmenu_multilineAllFolders" : "multilineAllFolders"
+  );
+  const isMultiline = Services.prefs.getBoolPref(
+    "mail.pane_config.multiline_all"
+  );
+  if (isMultiline) {
+    multiline.removeAttribute("checked");
+  } else {
+    multiline.setAttribute("checked", "true");
+  }
+  Services.prefs.setBoolPref("mail.pane_config.multiline_all", !isMultiline);
+}
+
 function MsgViewAllHeaders() {
   Services.prefs.setIntPref(
     "mail.show_headers",
     Ci.nsMimeHeaderDisplayTypes.AllHeaders
diff --git a/mail/base/content/messenger-menubar.inc.xhtml b/mail/base/content/messenger-menubar.inc.xhtml
--- a/mail/base/content/messenger-menubar.inc.xhtml
+++ b/mail/base/content/messenger-menubar.inc.xhtml
@@ -377,8 +377,13 @@
                   data-l10n-id="menu-view-layout-horizontal"
                   name="viewlayoutgroup"
                   command="cmd_viewHorizontalMailLayout"/>
         <menuseparator id="viewMenuAfterPaneVerticalSeparator"/>
+        <menuitem id="multilineAllFolders"
+                  type="checkbox"
+                  data-l10n-id="menu-multi-line-all-folders"
+                  oncommand="ToogleMultilineAll(false);"/>
+        <menuseparator id="viewMenuBeforeShowFolderPaneSeparator"/>
         <menuitem id="menu_showFolderPane" type="checkbox" label="&showFolderPaneCmd.label;"
                   accesskey="&showFolderPaneCmd.accesskey;" command="cmd_toggleFolderPane"/>
         <menuitem id="menu_toggleThreadPaneHeader"
                   type="checkbox"
diff --git a/mail/base/content/modules/ThreadPaneColumns.mjs b/mail/base/content/modules/ThreadPaneColumns.mjs
--- a/mail/base/content/modules/ThreadPaneColumns.mjs
+++ b/mail/base/content/modules/ThreadPaneColumns.mjs
@@ -77,31 +77,31 @@ const DEFAULT_COLUMNS = [
     icon: true,
     resizable: false,
   },
   {
+    id: "unreadButtonColHeader",
+    l10n: {
+      header: "threadpane-column-header-unread-button",
+      menuitem: "threadpane-column-label-unread-button",
+      cell: "threadpane-cell-read-status",
+    },
+    ordinal: 5,
+    sortKey: "byUnread",
+    icon: true,
+    resizable: false,
+  },
+  {
     id: "subjectCol",
     l10n: {
       header: "threadpane-column-header-subject",
       menuitem: "threadpane-column-label-subject",
       cell: "threadpane-cell-subject-title",
     },
-    ordinal: 5,
+    ordinal: 6,
     picker: false,
     sortKey: "bySubject",
   },
   {
-    id: "unreadButtonColHeader",
-    l10n: {
-      header: "threadpane-column-header-unread-button",
-      menuitem: "threadpane-column-label-unread-button",
-      cell: "threadpane-cell-read-status",
-    },
-    ordinal: 6,
-    sortKey: "byUnread",
-    icon: true,
-    resizable: false,
-  },
-  {
     id: "senderCol",
     l10n: {
       header: "threadpane-column-header-sender",
       menuitem: "threadpane-column-label-sender",
diff --git a/mail/base/content/widgets/treeview/thread-row.mjs b/mail/base/content/widgets/treeview/thread-row.mjs
--- a/mail/base/content/widgets/treeview/thread-row.mjs
+++ b/mail/base/content/widgets/treeview/thread-row.mjs
@@ -14,9 +14,10 @@ import { TreeViewTableRow } from "chrome
 class ThreadRow extends TreeViewTableRow {
   /**
    * The default height of the table row.
    */
-  static ROW_HEIGHT = 22;
+  static ROW_HEIGHT = 0; // Will be set via density.
+  static MULTILINE = false;
 
   connectedCallback() {
     if (this.hasConnected) {
       return;
@@ -28,8 +29,17 @@ class ThreadRow extends TreeViewTableRow
     this.classList.add("table-layout");
     this.appendChild(
       document.getElementById("threadPaneRowTemplate").content.cloneNode(true)
     );
+
+    const table = this.parentNode.parentNode;
+    if (ThreadRow.MULTILINE) {
+      this.setAttribute("multiline", "true");
+      table.setAttribute("multiline", "true");
+    } else {
+      this.removeAttribute("multiline");
+      table.removeAttribute("multiline");
+    }
   }
 
   _fillRow() {
     super._fillRow();
@@ -72,8 +82,9 @@ class ThreadRow extends TreeViewTableRow
     const isDummyRow = propertiesSet.has("dummy");
 
     this.dataset.properties = properties.value.trim();
 
+    let seenSubjectColumn = false;
     for (const column of window.threadPane.columns) {
       // Skip this column if it's hidden.
       if (column.hidden) {
         continue;
@@ -82,8 +93,9 @@ class ThreadRow extends TreeViewTableRow
       const textIndex = textColumns.indexOf(column.id);
 
       // Special case for the subject column.
       if (column.id == "subjectCol") {
+        seenSubjectColumn = true;
         const div = cell.querySelector(".subject-line");
 
         // Indent child message of this thread.
         div.style.setProperty(
@@ -124,8 +136,19 @@ class ThreadRow extends TreeViewTableRow
         document.l10n.setAttributes(cell, column.l10n.cell);
         continue;
       }
 
+      if (seenSubjectColumn && !column.hidden) {
+        seenSubjectColumn = false;
+        cell.setAttribute("alignmentcolumn", "true");
+        cell.style.setProperty(
+          "--thread-level",
+          window.gViewWrapper.showGroupedBySort ? 0 : threadLevel.value
+        );
+      } else {
+        cell.removeAttribute("alignmentcolumn");
+      }
+
       if (column.id == "threadCol") {
         let buttonL10nId, labelString;
         if (propertiesSet.has("ignore")) {
           buttonL10nId = "tree-list-view-row-ignored-thread-button";
@@ -233,8 +256,10 @@ class ThreadRow extends TreeViewTableRow
 
       if (textIndex >= 0) {
         if (isDummyRow) {
           cell.textContent = "";
+          cell.removeAttribute("alignmentcolumn");
+          cell.style.removeProperty("--thread-level");
           continue;
         }
         cell.textContent = cellTexts[textIndex];
         if (!column.custom) {
diff --git a/mail/components/customizableui/content/panelUI.inc.xhtml b/mail/components/customizableui/content/panelUI.inc.xhtml
--- a/mail/components/customizableui/content/panelUI.inc.xhtml
+++ b/mail/components/customizableui/content/panelUI.inc.xhtml
@@ -377,8 +377,14 @@
                        data-l10n-id="menu-view-layout-horizontal"
                        name="viewlayoutgroup"
                        command="cmd_viewHorizontalMailLayout"/>
         <toolbarseparator id="appmenu_viewMenuAfterPaneVerticalSeparator"/>
+        <toolbarbutton id="appmenu_multilineAllFolders"
+                       class="subviewbutton subviewbutton-iconic"
+                       type="checkbox"
+                       data-l10n-id="menu-multi-line-all-folders"
+                       oncommand="ToogleMultilineAll(true);"/>
+        <toolbarseparator id="appmenu_viewMenuBeforeShowFolderPaneSeparator"/>
         <toolbarbutton id="appmenu_showFolderPane"
                        class="subviewbutton subviewbutton-iconic"
                        type="checkbox"
                        closemenu="none"
diff --git a/mail/locales/en-US/messenger/menubar.ftl b/mail/locales/en-US/messenger/menubar.ftl
--- a/mail/locales/en-US/messenger/menubar.ftl
+++ b/mail/locales/en-US/messenger/menubar.ftl
@@ -122,8 +122,11 @@ menu-view-repair-text-encoding =
 menu-view-folders-toggle-header =
     .label = Folder Pane Header
     .accesskey = P
 
+menu-multi-line-all-folders =
+    .label = Multi-line View On All Folders
+
 ## View / Layout
 
 menu-view-layout-widethread =
     .label = Wide Thread View
diff --git a/mail/themes/shared/mail/threadPane.css b/mail/themes/shared/mail/threadPane.css
--- a/mail/themes/shared/mail/threadPane.css
+++ b/mail/themes/shared/mail/threadPane.css
@@ -403,9 +403,9 @@ tr[data-properties~="unread"].selected .
     box-sizing: border-box;
   }
 
   & .subject-line {
-    margin-inline-start: calc(16px * var(--thread-level));
+    margin-inline-start: calc(10px * var(--thread-level));
     pointer-events: none;
     /* To avoid vertical cut off of characters, JS in about3Pane's densityChange
       that needs to be updated if line height value is changed.*/
     line-height: 1.4;
@@ -434,8 +434,78 @@ tr[data-properties~="unread"].selected .
     background-image: url("chrome://messenger/skin/icons/search-row-end.svg");
   }
 }
 
+:root[uidensity="compact"] .table-layout[multiline="true"] .subject-line {
+  line-height: 20px;
+}
+
+/* Override rules from tree-listbox.css */
+.table-layout[multiline="true"] td div.thread-container,
+.table-layout[multiline="true"] td div.subject-line {
+  overflow: visible !important;
+}
+.table-layout[multiline="true"] td.subjectcol-column {
+  overflow: visible !important;
+}
+
+/* From bug 1845204: https://hg.mozilla.org/comm-central/rev/932332d4db39#l1.28 */
+tbody .table-layout[multiline="true"] td:not([colspan="2"]) button:not(.twisty) {
+  height: initial !important;
+}
+
+/* Indent "alignment column" as well. */
+.table-layout[multiline="true"] td[alignmentcolumn="true"] {
+  padding-left: calc(6px + 10px * var(--thread-level));
+}
+
+/* Remove centering of the td */
+.table-layout[multiline="true"] td {
+  vertical-align: top;
+}
+:root[uidensity="compact"] .table-layout[multiline="true"] td {
+  padding-top: 2px;
+}
+:root[uidensity="compact"] .table-layout[multiline="true"] td.subjectcol-column {
+  top: 7px;
+}
+.table-layout[multiline="true"] td {
+  padding-top: 6px;
+}
+.table-layout[multiline="true"] td.subjectcol-column {
+  top: 9px;
+}
+:root[uidensity="touch"] .table-layout[multiline="true"] td {
+  padding-top: 8px;
+}
+:root[uidensity="touch"] .table-layout[multiline="true"] td.subjectcol-column {
+  top: 10px;
+}
+
+/* Make the subject column header fixed width and remove the splitter */
+table[is="tree-view-table"][multiline="true"] #subjectCol {
+  width: 12px !important;
+  min-width: 12px !important;
+  outline: 1px solid var(--splitter-color);
+  border-right: 1px solid var(--splitter-color);
+}
+table[is="tree-view-table"][multiline="true"] #subjectColSplitter {
+  display: none !important;
+}
+table[is="tree-view-table"][multiline="true"] #subjectColButton {
+  content-visibility: hidden;
+  background-position-x: left 0;
+}
+tr[is="tree-view-table"][multiline="true"]:not([data-properties~="dummy"]) .subjectcol-column {
+  /* Italic is a matter of taste here. It can be removed in userChrome.css */
+  font-style: italic;
+}
+
+/* Centre the text of dummy rows (grouped view). */
+.table-layout[multiline="true"][data-properties~="dummy"] td.subjectcol-column {
+  top: 2px !important;
+}
+
 [is="tree-view-table-body"][role="treegrid"] .table-layout:not(.children) .subject-line {
   padding-inline-start: 22px;
 }
 
@@ -604,15 +674,30 @@ tr[data-properties~="junk"] :is(.tree-vi
   background-position-y: center;
   padding-inline-start: var(--tree-header-cell-padding) !important;
   -moz-context-properties: stroke, fill-opacity;
   stroke: currentColor;
-  fill-opacity: 0.3;
+  fill-opacity: 0.7;
 
   &:dir(rtl) {
     background-position-x: right 1px;
   }
 }
 
+.table-layout[multiline="true"] .correspondentcol-column {
+  background-position-x: calc(4px + 10px * var(--thread-level));
+  background-position-y: top 5px;
+}
+:root[uidensity="compact"] .table-layout[multiline="true"] .correspondentcol-column {
+  background-position-y: top 1px;
+}
+:root[uidensity="touch"] .table-layout[multiline="true"] .correspondentcol-column {
+  background-position-y: top 7px;
+}
+
+.table-layout[multiline="true"] td[alignmentcolumn="true"].correspondentcol-column {
+  padding-inline-start: calc(var(--tree-header-cell-padding) + 10px * var(--thread-level)) !important;
+}
+
 tr[data-properties~="outgoing"] .correspondentcol-column {
   background-image: var(--icon-nav-forward);
 
   &:dir(rtl) {
diff --git a/mail/themes/shared/mail/tree-listbox.css b/mail/themes/shared/mail/tree-listbox.css
--- a/mail/themes/shared/mail/tree-listbox.css
+++ b/mail/themes/shared/mail/tree-listbox.css
@@ -27,9 +27,9 @@
   --tree-header-cell-padding: 6px;
   --tree-header-cell-indent: 0;
 
   &[uidensity="compact"] {
-    --list-item-min-height: 18px;
+    --list-item-min-height: 20px;
     --list-item-buttons-size: 18px;
     --indicator-width: 3px;
     --indicator-inline-start: 1px;
     --table-header-cell-icon-width: 22px;
